## Numpy

Numpy is a library for the python programming languange, adding large multi-dimensional arrays and matrices, along with the large collections of the high
level mathematical functions to operate on these arrays
- very fast
- uses vectorization using brodcasting (without using loops)

### 1. Datatypes and Attributes
Numpy has only 1 datatype ndarray (n dimensional array)

In [5]:
import numpy as np

In [6]:
a1 = np.array([1,2,3])            # create a 1d array. It takes python list as parameters 
a1

array([1, 2, 3])

In [7]:
l1 = [4,5,6]
a1 = np.array(l1)

In [8]:
a1

array([4, 5, 6])

In [9]:
a2 = np.array([[1,2,3],            # create 2d array
               [4,5,6]])
a2

array([[1, 2, 3],
       [4, 5, 6]])

In [10]:
a3 = np.array([[[1, 2, 3],
                [4, 5, 6]],        #  create 3d array
               [[7, 8, 9],
                [10, 11, 12]],
               [[13, 14, 15],
                [16, 17, 18]]])

In [12]:
a3

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]])

In [14]:
a1.shape

(3,)

In [15]:
a2.shape

(2, 3)

In [23]:
a3.shape                          # check how many dimensions are there in array

(3, 2, 3)

In [24]:
a1.ndim , a2.ndim, a3.ndim        # check the dimensions of array

(1, 2, 3)

In [25]:
a1.dtype, a2.dtype, a3.dtype       # check the type of data stored in arrray

(dtype('int32'), dtype('int32'), dtype('int32'))

In [26]:
a1.size, a2.size, a3.size         # check the no of elements in the array

(3, 6, 18)

In [27]:
type(a1), type(a2), type(a3)      # check what is the type of the datastructure

(numpy.ndarray, numpy.ndarray, numpy.ndarray)

 ## 2. Creating Numpy Arrays

In [28]:
sample_array = np.array([1,2,3])
print(sample_array)

[1 2 3]


In [30]:
sample_array.dtype

dtype('int32')

In [31]:
ones = np.ones((2,3), dtype=float)    # gives a arrays with given shape and type, all filled with ones  

In [32]:
ones

array([[1., 1., 1.],
       [1., 1., 1.]])

In [33]:
print(ones.dtype)
print(type(ones))

float64
<class 'numpy.ndarray'>


In [34]:
zeros = np.zeros((3,3))   # arrays filled with zeros 

In [35]:
zeros

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [36]:
range_array = np.arange(0, 10, 2)

In [37]:
range_array

array([0, 2, 4, 6, 8])

In [38]:
random_array = np.random.randint(2, 9, size = (3,5))

In [39]:
random_array

array([[8, 3, 3, 3, 6],
       [5, 7, 7, 4, 6],
       [3, 3, 3, 8, 6]])

In [40]:
random_array2 = np.random.random((5,3))

In [41]:
random_array2

array([[0.54545034, 0.05990775, 0.87687031],
       [0.44172388, 0.88969806, 0.0912468 ],
       [0.97318032, 0.93690296, 0.11006074],
       [0.01938437, 0.64097538, 0.20994308],
       [0.74122262, 0.32436457, 0.61371477]])

In [42]:
random_array3 = np.random.rand(3,3)

In [43]:
random_array3

array([[0.33838705, 0.38958052, 0.81603963],
       [0.25403175, 0.5562384 , 0.95392395],
       [0.25304136, 0.03427383, 0.76161778]])

## 3. Viewing arrays and matrices

In [44]:
my_array = np.random.randint(2,9,size=(4,4))
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [45]:
np.unique(my_array)  # returns an array with all elements from the passed array only one time

array([2, 3, 4, 5, 6, 7, 8])

In [46]:
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [48]:
my_array[0], my_array[1], my_array[2], my_array[3]

(array([8, 7, 5, 5]),
 array([4, 4, 8, 2]),
 array([3, 8, 5, 5]),
 array([5, 2, 5, 6]))

In [49]:
my_array[0][0], my_array[0][1], my_array[0][2], my_array[0][3], 

(8, 7, 5, 5)

In [50]:
# We can also use the slicing in numpy arrays. Just like the python
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [51]:
my_array[0:2]

array([[8, 7, 5, 5],
       [4, 4, 8, 2]])

In [52]:
my_array[2:4]

array([[3, 8, 5, 5],
       [5, 2, 5, 6]])

In [55]:
my_array[0:len(my_array)]

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [57]:
my_array[0:4:2]

array([[8, 7, 5, 5],
       [3, 8, 5, 5]])

In [58]:
my_array[::-1]

array([[5, 2, 5, 6],
       [3, 8, 5, 5],
       [4, 4, 8, 2],
       [8, 7, 5, 5]])

In [59]:
my_array2 = np.random.randint(10, size = (2,3,4,5))
my_array2

array([[[[3, 0, 1, 9, 8],
         [4, 1, 5, 1, 2],
         [4, 4, 6, 3, 8],
         [9, 8, 1, 5, 9]],

        [[3, 2, 5, 8, 8],
         [3, 7, 0, 7, 3],
         [5, 2, 8, 3, 9],
         [6, 1, 7, 2, 4]],

        [[8, 3, 4, 8, 9],
         [4, 1, 8, 8, 5],
         [8, 2, 9, 0, 0],
         [9, 5, 6, 2, 9]]],


       [[[4, 2, 1, 3, 8],
         [2, 2, 7, 1, 9],
         [6, 3, 4, 1, 4],
         [9, 4, 5, 1, 5]],

        [[1, 2, 1, 1, 2],
         [5, 6, 2, 4, 1],
         [6, 6, 6, 9, 3],
         [1, 1, 3, 4, 2]],

        [[9, 4, 3, 3, 1],
         [3, 4, 9, 9, 0],
         [9, 8, 2, 7, 4],
         [3, 9, 9, 7, 7]]]])

In [60]:
# get the first 4 numbers of the inner most arrays 
my_array2[:, :, :, :1]

array([[[[3],
         [4],
         [4],
         [9]],

        [[3],
         [3],
         [5],
         [6]],

        [[8],
         [4],
         [8],
         [9]]],


       [[[4],
         [2],
         [6],
         [9]],

        [[1],
         [5],
         [6],
         [1]],

        [[9],
         [3],
         [9],
         [3]]]])

 ## 4. Manipulating and Comparing Arrays

### Arithmetic

In [61]:
ones =  np.ones(3)

In [62]:
ones 

array([1., 1., 1.])

In [74]:
a1 = np.array([1,2,3])
a1

array([1, 2, 3])

In [75]:
a1 + ones

array([2., 3., 4.])

In [76]:
a1 - ones

array([0., 1., 2.])

In [77]:
a1 * ones 

array([1., 2., 3.])

In [78]:
a2 = np.array([[1,2,3], [4,5,6]])
a2

array([[1, 2, 3],
       [4, 5, 6]])

In [79]:
a1 * a2

array([[ 1,  4,  9],
       [ 4, 10, 18]])

In [80]:
a3 = np.array([[7,8,9], [11,12,13]])
a3

array([[ 7,  8,  9],
       [11, 12, 13]])

In [81]:
a2 * a3     # mmultiply adjacent elements 

array([[ 7, 16, 27],
       [44, 60, 78]])

In [82]:
a1 * 2

array([2, 4, 6])

In [83]:
a1 / ones 

array([1., 2., 3.])

In [84]:
a2 / a1

array([[1. , 1. , 1. ],
       [4. , 2.5, 2. ]])

In [85]:
a2 // a1 # floor division

array([[1, 1, 1],
       [4, 2, 2]], dtype=int32)

In [86]:
a1 ** 2

array([1, 4, 9], dtype=int32)

In [87]:
a2 ** a1

array([[  1,   4,  27],
       [  4,  25, 216]], dtype=int32)

In [89]:
np.square(a2)

array([[ 1,  4,  9],
       [16, 25, 36]], dtype=int32)

In [95]:
a1

array([1, 2, 3])

In [96]:
ones 

array([1., 1., 1.])

In [97]:
np.add(a1, ones )

array([2., 3., 4.])

In [99]:
np.mod(a2,2) # return remainder after division

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [100]:
a2 % 2

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [103]:
a2 / 2

array([[0.5, 1. , 1.5],
       [2. , 2.5, 3. ]])

In [104]:
a2%2

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [105]:
np.log(a1)

array([0.        , 0.69314718, 1.09861229])

In [106]:
np.exp(a2)

array([[  2.71828183,   7.3890561 ,  20.08553692],
       [ 54.59815003, 148.4131591 , 403.42879349]])

## Aggregation
Performing the same operation on a number of things 

In [109]:
ls1 = [1,2,3]
type(ls1)

list

In [110]:
sum(ls1)

6

In [111]:
a1 = np.array([1,2,3])
sum(a1)

6

In [112]:
np.sum(a1)

6

- It is advised to perform pythons methods on python datatypes and use numpy's methods on the Numpy arrays

In [117]:
# Create a massive numpy array
massive_array = np.random.random(100000)
massive_array

array([0.90531317, 0.09193317, 0.96397208, ..., 0.47429839, 0.11038701,
       0.289257  ])

In [118]:
massive_array.size

100000

In [119]:
massive_array.dtype

dtype('float64')

In [120]:
type(massive_array)

numpy.ndarray

In [121]:
%timeit sum(massive_array)   #Python's sum method
%timeit np.sum(massive_array) #Numpy sum method

10.4 ms ± 584 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
35.5 µs ± 1.53 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


`%timeit` is the jupyter magic function. It give the time taken by the line to execute.

##### As we can see from the above calculation python `sum()` method take 10.4 milliseconds  while Numpy `np.sum()` method take 35.5 microseconds to perform the sum of 100000 entries.
Pyhton method is 500+ times slower than numpy function. 
<br>
Therefore numpy is great library for ML